### PR TITLE
Improved healthcheck

### PR DESCRIPTION
**Pull request by andrewkononenko**

### CODE REVIEW

File: MailManager/src/main/java/inc/softserve/health/MailManagerHealthCheck.java [hunk 1]
```diff
@@ -31,32 +34,49 @@ public MailManagerHealthCheck(@ESPConnectorAdmin String ESPConnectorAdminUrl,
...
+        ExecutorService executor = Executors.newFixedThreadPool(3);
+        FutureTask<Boolean> espConnectorCheck = new FutureTask(() -> isEspConnectorHealthy());
+        FutureTask<Boolean> mongoCheck = new FutureTask(() -> isMongoHealthy());
 
+        FutureTask<Boolean> commonHealth = new FutureTask(() -> {
+            executor.execute(espConnectorCheck);
+            executor.execute(mongoCheck);
+            espConnectorCheck.get();
+            mongoCheck.get();
+            return espConnectorCheck.get() && mongoCheck.get();
```
Comment (svoronin) [inline MailManager/src/main/java/inc/softserve/health/MailManagerHealthCheck.java:44]: Why do we need to get those results twice?

File: MailManager/src/main/java/inc/softserve/health/MailManagerHealthCheck.java [hunk 2]
```diff
@@ -31,32 +34,49 @@ public MailManagerHealthCheck(@ESPConnectorAdmin String ESPConnectorAdminUrl,
...
         } catch (Exception e) {
-            e.printStackTrace();
+            healthy = false;
         }
 
-        if (!espConnectorHealthy) {
-            unhealthyMessage = unhealthyMessage.append("ESPConnector is unavailable!");
+        if (healthy) {
+            return Result.healthy();
         }
```
Comment (svoronin) [inline MailManager/src/main/java/inc/softserve/health/MailManagerHealthCheck.java]: It look like you can get rid of `healthy` variable and just return the result.

File: MailManager/src/main/java/inc/softserve/health/MailManagerHealthCheck.java [hunk 3]
```diff
@@ -31,32 +34,47 @@ public MailManagerHealthCheck(@ESPConnectorAdmin String ESPConnectorAdminUrl,
...
+        return Result.unhealthy("Service unavailable!");
+    }
 
+    boolean isEspConnectorHealthy(){
```
Comment (svoronin) [inline MailManager/src/main/java/inc/softserve/health/MailManagerHealthCheck.java]: Keep those methods private please

File: MailManager/src/main/java/inc/softserve/health/MailManagerHealthCheck.java [hunk 4]
```diff
@@ -31,32 +34,47 @@ public MailManagerHealthCheck(@ESPConnectorAdmin String ESPConnectorAdminUrl,
     }
 
     protected Result check() throws Exception {
-        boolean espConnectorHealthy = false;
-        StringBuilder unhealthyMessage = new StringBuilder("");
+        ExecutorService executor = Executors.newFixedThreadPool(3);
+        FutureTask<Boolean> espConnectorCheck = new FutureTask(() -> isEspConnectorHealthy());
+        FutureTask<Boolean> mongoCheck = new FutureTask(() -> isMongoHealthy());
```
Comment (svoronin) [inline MailManager/src/main/java/inc/softserve/health/MailManagerHealthCheck.java]: Please use method references here, as discussed.

File: MailManager/src/main/java/inc/softserve/health/MailManagerHealthCheck.java [hunk 5]
```diff
@@ -31,32 +34,47 @@ public MailManagerHealthCheck(@ESPConnectorAdmin String ESPConnectorAdminUrl,
...
 
+        FutureTask<Boolean> commonHealth = new FutureTask(() -> {
+            executor.execute(espConnectorCheck);
+            executor.execute(mongoCheck);
+            return espConnectorCheck.get() && mongoCheck.get();
+        });
```
Comment (svoronin) [inline MailManager/src/main/java/inc/softserve/health/MailManagerHealthCheck.java:45]: My compiler complains about all those three future declatarions. Could you please fix related issue please?

File: MailManager/src/main/java/inc/softserve/health/MailManagerHealthCheck.java [hunk 6]
```diff
@@ -30,33 +33,42 @@ public MailManagerHealthCheck(@ESPConnectorAdmin String ESPConnectorAdminUrl,
...
+
+        try {
+            executor.execute(commonHealth);
+            return (commonHealth.get(500L, TimeUnit.MILLISECONDS)) ? Result.healthy() : Result.unhealthy("Service unavailable!");
```
Comment (svoronin) [inline MailManager/src/main/java/inc/softserve/health/MailManagerHealthCheck.java]: That is preferable to split in three lines: condition, true and false statements.

File: MailManager/src/main/java/inc/softserve/health/MailManagerHealthCheck.java [hunk 7]
```diff
@@ -35,18 +35,20 @@ public MailManagerHealthCheck(@ESPConnectorAdmin String ESPConnectorAdminUrl,
 
     protected Result check() {
```
Comment (svoronin) [inline MailManager/src/main/java/inc/softserve/health/MailManagerHealthCheck.java:36]: This method is now awesome. Are you up to do it using `CompletableFuture`? :smiley:

Comment (andrewkononenko) [inline MailManager/src/main/java/inc/softserve/health/MailManagerHealthCheck.java:36]: In future releases.

**Comment by svoronin** [comment]
As you wish :wink:

**Pull request merged**